pricegraph/src/graph/path.rs

```diff
@@ -1,46 +1,84 @@
 //! Utilities for finding paths from predecessor vectors.
 
-use petgraph::graph::NodeIndex;
+use petgraph::visit::NodeIndexable;
 
-/// Finds a cycle by searching from the provided `search` node and returns a
-/// vector representing a path along the cycle. This method returns a path,
-/// which means that the first and last nodes will always be the same.
+#[derive(Debug, PartialEq)]
+/// A path of nodes connected by a (directed) edge.
+pub struct Path<N>(pub Vec<N>);
+
+impl<N> From<NegativeCycle<N>> for Path<N> {
+    fn from(cycle: NegativeCycle<N>) -> Self {
+        Path(cycle.0)
+    }
+}
+
+#[derive(Debug)]
+/// An ordered collection of nodes that form a cycle of negative weight.
+/// The first node of the cycle coincides with the last.
+pub struct NegativeCycle<N>(pub Vec<N>);
+
+impl<N: PartialEq> NegativeCycle<N> {
+    /// Returns the negative cycle changing its starting and terminating
+    /// node to be the given node. If the given node is not part of the
+    /// cycle, it returns an error containing the original cycle.
+    pub fn with_starting_node(self, start: N) -> Result<NegativeCycle<N>, NegativeCycle<N>> {
+        let mut cycle = Vec::with_capacity(self.0.len());
+        let mut cycle_end = Vec::with_capacity(self.0.len());
+        let mut self_iterator = self.0.into_iter();
+        while let Some(node) = self_iterator.next() {
+            if node == start {
+                cycle.push(node);
+                cycle.append(&mut self_iterator.collect());
+                cycle.pop(); // the last element of self.0 is identical to the first
+                cycle.append(&mut cycle_end);
+                cycle.push(start);
+                return Ok(NegativeCycle(cycle));
+            } else {
+                cycle_end.push(node);
+            }
+        }
+        Err(NegativeCycle(cycle_end))
+    }
```

Comment:
I think this can be written more efficiently:
```suggestion
    pub fn with_starting_node(mut self, start: N) -> Result<NegativeCycle<N>, NegativeCycle<N>> {
        match self.0.iter().position(|i| *i == start) {
            None => Err(self),
            Some(pos) if pos == 0 => Ok(self),
            Some(pos) => {
                self.0.pop();
                self.0.rotate_right(pos);
                debug_assert!(self.0[0] == start);
                self.0.push(start);
                Ok(self)
            }
        }
    }
```

Reply:
Nice!

Reply:
Or maybe use the rotate method on slice (https://doc.rust-lang.org/std/primitive.slice.html#method.rotate_left)?